solidity/contracts/AbstractBondedECDSAKeep.sol

```diff
@@ -0,0 +1,643 @@
+    /// this keep.
+    /// @return true if the keep is closed, false otherwise.
+    function isClosed() public view returns (bool) {
+        return status == Status.Closed;
+    }
+
+    /// @notice Returns true if the keep has been terminated.
+    /// Keep is terminated when bonds are seized and members no longer support
+    /// this keep.
+    /// @return true if the keep has been terminated, false otherwise.
+    function isTerminated() public view returns (bool) {
+        return status == Status.Terminated;
+    }
+
+    /// @notice Returns members of the keep.
+    /// @return List of the keep members' addresses.
+    function getMembers() public view returns (address[] memory) {
+        return members;
+    }
+
+    /// @notice Checks a fraud proof for a valid signature from this keep that was
+    /// not first approved via a call to sign.
+    /// @dev The function expects the signed digest to be calculated as a sha256 hash
+    /// of the preimage: `sha256(_preimage))`. The digest is verified against the
+    /// preimage to ensure the security of the ECDSA protocol. Verifying just the
+    /// signature and the digest is not enough and leaves the possibility of the
+    /// the existential forgery. If digest and preimage verification fails the
+    /// function reverts.
+    /// Reverts if a public key has not been set for the keep yet.
+    /// @param _v Signature's header byte: `27 + recoveryID`.
+    /// @param _r R part of ECDSA signature.
+    /// @param _s S part of ECDSA signature.
+    /// @param _signedDigest Digest for the provided signature. Result of hashing
+    /// the preimage with sha256.
+    /// @param _preimage Preimage of the hashed message.
+    /// @return True if fraud, false otherwise.
+    function checkSignatureFraud(
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s,
+        bytes32 _signedDigest,
+        bytes memory _preimage
+    ) public view returns (bool _isFraud) {
+        require(publicKey.length != 0, "Public key was not set yet");
+
+        bytes32 calculatedDigest = sha256(_preimage);
+        require(
+            _signedDigest == calculatedDigest,
+            "Signed digest does not match sha256 hash of the preimage"
+        );
+
+        bool isSignatureValid = publicKeyToAddress(publicKey) ==
+            ecrecover(_signedDigest, _v, _r, _s);
+
+        // Check if the signature is valid but was not requested.
+        return isSignatureValid && digests[_signedDigest] == 0;
+    }
+
+    /// @notice Initialization function.
+    /// @dev We use clone factory to create new keep. That is why this contract
+    /// doesn't have a constructor. We provide keep parameters for each instance
+    /// function after cloning instances from the master contract.
+    /// @param _owner Address of the keep owner.
+    /// @param _members Addresses of the keep members.
+    /// @param _honestThreshold Minimum number of honest keep members.
+    function initializeBondedECDSAKeep(
```

Comment:
What do you think about naming it just `initialize`? It's kinda weird that `initialize` in `BondedECDSAKeep` calls `initializeBondedECDSAKeep` that is in `AbstractBondedECDSAKeep`.